### PR TITLE
[Issue #42] Add control and test for token authentication

### PR DESCRIPTION
**Pull request by khermano**
Issue: https://github.com/xstefank/tyr/issues/42

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -81,11 +92,27 @@
             <version>2.22.2</version>
             <scope>test</scope>
         </dependency>
+
+        <dependency>
+            <groupId>org.jboss.arquillian.junit</groupId>
+            <artifactId>arquillian-junit-container</artifactId>
+            <version>1.4.0.Final</version>
```
Comment (xstefank) [inline pom.xml]: cannot this version be taken from imported arquillian bom?

File: pom.xml [hunk 2]
```diff
@@ -81,11 +92,27 @@
...
 
     <build>
         <finalName>tyr</finalName>
         <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <version>3.8.0</version>
+                <configuration>
+                    <source>1.8</source>
```
Comment (xstefank) [inline pom.xml]: you can do the same with <maven.compiler.source> and <maven.compiler.target> properties. Please move this configuration into properties

File: pom.xml [hunk 3]
```diff
@@ -137,4 +164,85 @@
             </plugin>
         </plugins>
     </build>
+
+    <profiles>
+        <profile>
+            <id>wildfly-managed</id>
```
Comment (xstefank) [inline pom.xml]: maybe different name? what this really does is integration tests right?

File: pom.xml [hunk 4]
```diff
@@ -137,4 +164,85 @@
...
+                <plugins>
+                    <plugin>
+                        <artifactId>maven-surefire-plugin</artifactId>
+                        <version>2.16</version>
+                        <configuration>
+                            <trimStackTrace>false</trimStackTrace>
+                            <systemPropertyVariables>
+                                <arquillian.launch>wildfly-managed</arquillian.launch>
+                                <java.util.logging.manager>org.jboss.logmanager.LogManager</java.util.logging.manager>
+                                <jboss.home>${project.basedir}/target/wildfly-16.0.0.Final</jboss.home>
```
Comment (xstefank) [inline pom.xml]: please make the version of WFLY configurable in properties

File: pom.xml [hunk 5]
```diff
@@ -137,4 +164,85 @@
...
+                                <goals>
+                                    <goal>integration-test</goal>
+                                    <goal>verify</goal>
+                                </goals>
+                                <configuration>
+                                    <systemPropertyVariables>
+                                        <arquillian.launch>wildfly-managed</arquillian.launch>
+                                        <java.util.logging.manager>org.jboss.logmanager.LogManager</java.util.logging.manager>
+                                        <jboss.home>${project.basedir}/target/wildfly-16.0.0.Final</jboss.home>
+                                        <module.path>${project.basedir}/target/wildfly-16.0.0.Final/modules</module.path>
```
Comment (xstefank) [inline pom.xml]: this is same configuration as in surefire plugin. Is this necessary? If so why?

File: src/test/java/org/xstefank/api/AuthorizationTestIT.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+import org.junit.runner.RunWith;
+import org.jboss.arquillian.junit.Arquillian;
+import org.xstefank.JaxRsApplication;
+
+import javax.ws.rs.core.UriBuilder;
+import java.net.URL;
+
+@RunWith(Arquillian.class)
+@RunAsClient
+public class AuthorizationTestIT {
```
Comment (xstefank) [inline src/test/java/org/xstefank/api/AuthorizationTestIT.java]: Is it authorization or authentication?

File: src/test/java/org/xstefank/api/AuthorizationTestIT.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    @ArquillianResource
+    private URL deploymentUrl;
+
+    @Test (expected=IllegalArgumentException.class)
+    public void testWrongAuthentication() {
+        GitHubAPI.getJsonFromUri(UriBuilder.fromPath(deploymentUrl.toString()).path("/test/fakeGithub_ReturnStatusUnauthorized").build());
+    }
+
+    @Test (expected=IllegalArgumentException.class)
+    public void testWrongAuthentication2() {
```
Comment (xstefank) [inline src/test/java/org/xstefank/api/AuthorizationTestIT.java]: please rename this

File: src/test/java/org/xstefank/api/AuthorizationTestIT.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+                .create(WebArchive.class, "test.war")
+                .addClasses(FakeGitHub.class, JaxRsApplication.class);
+    }
+
+    @ArquillianResource
+    private URL deploymentUrl;
+
+    @Test (expected=IllegalArgumentException.class)
+    public void testWrongAuthentication() {
+        GitHubAPI.getJsonFromUri(UriBuilder.fromPath(deploymentUrl.toString()).path("/test/fakeGithub_ReturnStatusUnauthorized").build());
```
Comment (xstefank) [inline src/test/java/org/xstefank/api/AuthorizationTestIT.java]: there is also fromURI method in UriBuilder

File: src/test/java/org/xstefank/api/FakeGitHub.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
+package org.xstefank.api;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.core.Response;
+
+@Path("/fakeGitHub_")
```
Comment (xstefank) [inline src/test/java/org/xstefank/api/FakeGitHub.java]: just curiosity -- why underscore at the end?

File: pom.xml (hunk 1 above)
Comment (khermano) [reply pom.xml]: unfortunately no... at least what I know...http://arquillian.org/modules/core-platform/#artifacts

File: pom.xml [hunk 10]
```diff
@@ -81,11 +93,28 @@
...
     </dependencies>
 
     <build>
         <finalName>tyr</finalName>
         <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <version>3.8.0</version>
+            </plugin>
```
Comment (xstefank) [inline pom.xml]: is compiler plugin definition still required if there is no additional configuration?

File: pom.xml [hunk 11]
```diff
@@ -137,4 +166,65 @@
...
+                                <goals>
+                                    <goal>integration-test</goal>
+                                    <goal>verify</goal>
+                                </goals>
+                                <configuration>
+                                    <systemPropertyVariables>
+                                        <arquillian.launch>wildfly-managed</arquillian.launch>
+                                        <java.util.logging.manager>org.jboss.logmanager.LogManager</java.util.logging.manager>
+                                        <jboss.home>${project.basedir}/target/wildfly-${version.wildfly}</jboss.home>
+                                        <module.path>${project.basedir}/target/wildfly-${version.wildfly}/modules</module.path>
```
Comment (xstefank) [inline pom.xml]: there is also project.build.directory

Comment (xstefank) [reply pom.xml]: and I believe you can reuse the path defined on the previous line, if not directly than you can define it as global property

File: src/test/java/org/xstefank/api/AuthenticationTestIT.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    @Test (expected=IllegalArgumentException.class)
+    public void testAuthenticationStatusInternalServerError() {
+        Assert.assertEquals(Response.Status.INTERNAL_SERVER_ERROR.getStatusCode(), getEndPointResponse(RETURN_STATUS_INTERNAL_SERVER_ERROR_PATH).getStatus());
+        testAuthentication(RETURN_STATUS_INTERNAL_SERVER_ERROR_PATH);
+    }
+
+    private Response getEndPointResponse(String endpointPath) {
+        Client restClient = ClientBuilder.newClient();
```
Comment (xstefank) [inline src/test/java/org/xstefank/api/AuthenticationTestIT.java]: client instance can be reused, and should be closed after tests - try before/after class annotations

File: src/test/java/org/xstefank/api/AuthenticationTestIT.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Test (expected=IllegalArgumentException.class)
+    public void testAuthenticationStatusUnauthorized() {
+        Assert.assertEquals(Response.Status.UNAUTHORIZED.getStatusCode(), getEndPointResponse(RETURN_STATUS_UNAUTHORIZED_PATH).getStatus());
+        testAuthentication(RETURN_STATUS_UNAUTHORIZED_PATH);
+    }
+
+    @Test (expected=IllegalArgumentException.class)
+    public void testAuthenticationStatusInternalServerError() {
+        Assert.assertEquals(Response.Status.INTERNAL_SERVER_ERROR.getStatusCode(), getEndPointResponse(RETURN_STATUS_INTERNAL_SERVER_ERROR_PATH).getStatus());
```
Comment (xstefank) [inline src/test/java/org/xstefank/api/AuthenticationTestIT.java]: I'm not sure if this is the right approach as you are basically making same rest call twice if I'm not mistaken. But ok, if you want to verify that the paths are right I will merge this

Comment (khermano) [reply src/test/java/org/xstefank/api/AuthenticationTestIT.java]: Thanks :)

**Pull request merged**